Single util method for all get config classes.

sancar opened the pull request:
Main purpose is to remove code duplication.
Also some of the getConfig methods were not using same logic
like client.getReliableTopic. It is fixed with this pr.

fixes https://github.com/hazelcast/hazelcast/issues/13880

sancar commented:
I could not find solution for Map and Nearcache. 
They have an extra step that does not fit into this solution.
`initDefaultMaxSizeForOnHeapMaps`
see https://github.com/hazelcast/hazelcast/blob/486a38ad635025b08eb0ecae36c38c13346cbde5/hazelcast/src/main/java/com/hazelcast/config/Config.java#L461

sancar commented on hazelcast-client/src/main/java/com/hazelcast/client/config/ClientReliableTopicConfig.java:
```diff
@@ -69,9 +82,12 @@ public String getName() {
+     * @return the updated ClientReliableTopicConfig
      */
-    public void setName(String name) {
+    public ClientReliableTopicConfig setName(String name) {
```
I was just writing a comment about this because I wasn't sure.

sancar replied:
I can get around this by using the method with nameSetter for all. 
```
 public static <T> T getConfig(ConfigPatternMatcher configPatternMatcher,
                                  Map<String, T> configs, String name,
                                  Class clazz, BiConsumer<T, String> nameSetter) 
```
Let me change it that way, so that we can discuss further if we want this prd or not. 

My main purpose is to make the behaviour of `client.getReliableTopic` aligned with all others. 
This prd is to standardize the way, so that we won't do the similar mistake again.

sancar commented on hazelcast/src/main/java/com/hazelcast/config/EventJournalConfig.java:
```diff
@@ -51,7 +51,7 @@
     private String mapName;
     private String cacheName;
-    private boolean enabled = true;
+    private boolean enabled = false;
```
I had to make if false to use ConfigUtils.getConfig method for `EventJournalConfig.java`
Why we have this enabled by default ?

mmedenjak replied:
This was introduced for use in jet and the reasoning for enabling it by default was - if you add it to the config, you are intending to use it. You don't need an additional step for enabling it. Same as for `MapConfig` - if you add it, you're using it.
In any case, this isn't going to work as it is not backwards compatible.

sancar replied:
I will revert the part with EventJournalConfig.

sancar replied:
The reason I have changed this is a test failure , `ClientMapNearCacheSerializationCountTest`
It seems that EventJournalConfig is added by default to Map. May be I can avoid adding `EventJournalConfig` instead of changing the default.

sancar replied:
I have decided to revert get*EventJournal methods completely, because behaviour is documented in their javadoc clearly 
```
If there is no default config as well, it will create one and disable the event journal by default.
```

mmedenjak replied:
Yes, the behaviour is here: https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/config/Config.java#L3119-L3122

pveentjer commented on hazelcast-client/src/test/java/com/hazelcast/client/map/impl/nearcache/ClientMapNearCacheSerializationCountTest.java:
```diff
@@ -100,32 +98,32 @@
-
-                {GET_ALL, newInt(1, 1, 1), newInt(0, 1, 1), newInt(1, 1, 1), newInt(1, 1, 1), OBJECT, null, null, null},
-                {GET_ALL, newInt(1, 1, 1), newInt(0, 1, 1), newInt(1, 1, 0), newInt(1, 1, 0), OBJECT, OBJECT, false, true},
+//                {GET, newInt(1, 1, 1), newInt(0, 0, 0), newInt(1, 0, 0), newInt(0, 1, 1), BINARY, null, null, null},
```
Do we want to keep the commented stuff?

sancar replied:
My mistake, this is a leftover from debugging an issue. Fixing it right now.

ahmetmircik commented on hazelcast/src/test/java/com/hazelcast/internal/config/ConfigUtilsTest.java:
```diff
@@ -24,12 +27,72 @@
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static groovy.util.GroovyTestCase.assertEquals;
```
?

sancar commented:
@pveentjer @ahmetmircik @mmedenjak I have addressed the comments. What is your thoughts on this ?  To fix  #13880 there was a much simpler way. I could have just copy and paste the getConfig methods as it is done for now. 
I am hoping that this will reduce the errors, but there is already exceptions: like getMapConfig, getNearCacheConfig, or EventJournalConfig.

ahmetmircik commented:
regarding this comment: https://github.com/hazelcast/hazelcast/pull/14014#issuecomment-432267915

We can move `initDefaultMaxSizeForOnHeapMaps` to map-proxy creation on OS and EE.

ahmetmircik commented:
For the configs that cannot extend `NamedConfig`, we can create a 4.0 issue to log them for the future.

sancar commented:
Issue related to named config is created here and assigned to 4.0 
https://github.com/hazelcast/hazelcast/issues/14144

sancar commented:
We have worked on this with @ahmetmircik . MapConfig and EventJournalConfig now also use new utility method. 
@pveentjer @mmedenjak  Can I get OK from you guys also ?

mmedenjak commented on hazelcast/src/main/java/com/hazelcast/config/Config.java:
```diff
@@ -3257,19 +2948,7 @@ public MerkleTreeConfig findMapMerkleTreeConfig(String name) {
-        config = new MerkleTreeConfig(defConfig).setMapName(name);
-        addMerkleTreeConfig(config);
-        return config;
+        return ConfigUtils.getConfig(configPatternMatcher, mapMerkleTreeConfigs, name, MerkleTreeConfig.class);
```
The merkle tree config is similar to the event journal config. It needs to be disabled if the name is "default".

sancar replied:
I did not notice that. I will fix it.

mmedenjak commented on hazelcast/src/main/java/com/hazelcast/config/MerkleTreeConfig.java:
```diff
@@ -230,6 +230,16 @@ public int hashCode() {
     }
 
+    @Override
+    public MerkleTreeConfig setName(String name) {
```
This isn't going to be nice when/if we add merkle tree support for cache as well. The class will then be similar to `EventJournalConfig`, having `mapName` and `cacheName`.

sancar replied:
Then let me revert this file and I can use similar method that I have used with `EvenJournalConfig` to set name.

ahmetmircik commented:
@sancar do you plan to fix #13880 for 3.11.1?

sancar commented:
@ahmetmircik 
Yes, and here is the backport 
https://github.com/hazelcast/hazelcast/pull/14163
